### PR TITLE
AArch64 implementation of PS FMA and ADDSUB, and some PD load/store/convert

### PR DESCRIPTION
**Pull request by JishinMaster**
Hi,

I have implemented, using the AARCH64 Neon intrinsics, the following functions : 
_mm_cvtpd_ps 
_mm_cvtps_pd
 _mm_castpd_si128
 _mm_fmadd_ps,
_mm_addsub_ps
_mm_load_pd
 _mm_loadu_pd
 _mm_store_pd
 _mm_storeu_pd

I have got some more new additions which comes from my simd_utils library, but it seemed too much for one pull request.

Please let me know if you need modifications to this pull request.

### CODE REVIEW

**Comment by jserv** [comment]
@JishinMaster, Thanks for contributing. Can you use `clang-format` to indent? Also, A32 specific implementations are missing.

File: sse2neon.h [hunk 1]
```diff
@@ -1219,8 +1271,8 @@ FORCE_INLINE __m128i _mm_shuffle_epi8(__m128i a, __m128i b)
     __asm__ __volatile__(
         "vtbl.8  %e[ret], {%e[tbl], %f[tbl]}, %e[idx]\n"
         "vtbl.8  %f[ret], {%e[tbl], %f[tbl]}, %f[idx]\n"
-        : [ret] "=&w"(ret)
-        : [tbl] "w"(tbl), [idx] "w"(idx_masked));
+        : [ ret ] "=&w"(ret)
```
Comment (jserv) [inline sse2neon.h:1274]: clang-format-10 is known to a bit weird for inline assembly. You can use [clang-format-6](https://releases.llvm.org/6.0.1/tools/clang/docs/ClangFormat.html).

**Comment by JishinMaster** [comment]
I have applied clang-format (make indent) with clang-format-10, and checked with the check-format.sh script.
Travis uses clang-format-6 and still complains about indentation.

I have added the armv7 neon equivalent where it made some sense (fmadd_ps and addsub_ps).
I do no think it would be relevant to manually implement other float64 functions using the armv7 neon instructions, since it was not native.
We would be better of using vfpv4 or even plain C, or have a define flag which would turn on float64 on AArch64 CPUs. 
What do you think?

**Comment by jserv** [comment]
> I have added the armv7 neon equivalent where it made some sense (fmadd_ps and addsub_ps).
> I do no think it would be relevant to manually implement other float64 functions using the armv7 neon instructions, since it was not native.
> We would be better of using vfpv4 or even plain C, or have a define flag which would turn on float64 on AArch64 CPUs.
> What do you think?

If some intrinsic is unlikely to be implemented in A32-only, C implementation can fit. We don't have to subject single pull request with full optimizations

**Comment by jserv** [comment]
Related: #70

File: sse2neon.h [hunk 2]
```diff
@@ -2390,6 +2442,41 @@ FORCE_INLINE __m128i _mm_maddubs_epi16(__m128i _a, __m128i _b)
...
+#endif
+}
+
+
+
+// Alternatively add and subtract packed single-precision (32-bit)
+// floating-point elements in a to/from packed elements in b, and store the
+// results in dst.
+//
+// The SSE3 flag is required for this function to be available on X86_64
```
Comment (jserv) [inline sse2neon.h:2470]: The comment about SSE3 is not necessary to SSE2NEON.

File: sse2neon.h [hunk 3]
```diff
@@ -2390,6 +2442,41 @@ FORCE_INLINE __m128i _mm_maddubs_epi16(__m128i _a, __m128i _b)
...
+// https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=addsub_ps&expand=2549,223
+FORCE_INLINE __m128 _mm_addsub_ps(__m128 a, __m128 b)
+{
+    __m128 mask = {-1.0f, 1.0f, -1.0f, 1.0f};
+    return _mm_fmadd_ps(b, mask, a);
+}
+
```
Comment (jserv) [inline sse2neon.h:2477]: Change 3 blank lines to single for consistent style.

File: sse2neon.h [hunk 4]
```diff
@@ -2390,6 +2442,41 @@ FORCE_INLINE __m128i _mm_maddubs_epi16(__m128i _a, __m128i _b)
...
+{
+#if defined(__aarch64__)
+    return vreinterpretq_m128_f32(vfmaq_f32(vreinterpretq_f32_m128(c),
+                                            vreinterpretq_f32_m128(b),
+                                            vreinterpretq_f32_m128(a)));
+#else
+    return _mm_add_ps(_mm_mul_ps(a, b), c);
+#endif
+}
+
```
Comment (jserv) [inline sse2neon.h:2463]: Ditto. Style.

File: sse2neon.h [hunk 5]
```diff
@@ -2390,6 +2442,41 @@ FORCE_INLINE __m128i _mm_maddubs_epi16(__m128i _a, __m128i _b)
     return vreinterpretq_m128i_s16(vqaddq_s16(prod1, prod2));
 }
 
+
```
Comment (jserv) [inline sse2neon.h:2445]: Ditto. Style.

**Comment by jserv** [comment]
Thank @JishinMaster for contributing!

File: tests/impl.cpp [hunk 6]
```diff
@@ -974,6 +980,18 @@ bool test_mm_load_sd(const double *p)
     return validateDouble(a, p[0], 0);
 }
 
+bool test_mm_load_pd(const double *p)
+{
+    __m128d a = _mm_load_pd(p);
+    return validateDouble(a, p[0], p[1]);
+}
+
+bool test_mm_loadu_pd(const double *p)
```
Comment (danlark1) [inline tests/impl.cpp:989]: I don't see that this function to be called in the tests

File: sse2neon.h [hunk 7]
```diff
@@ -2390,6 +2439,33 @@ FORCE_INLINE __m128i _mm_maddubs_epi16(__m128i _a, __m128i _b)
     return vreinterpretq_m128i_s16(vqaddq_s16(prod1, prod2));
 }
 
+// Computes the fused multiple add product of 32-bit floating point numbers.
+//
+// Return Value
+// Multiplies A and B, and adds C to the temporary result before returning it.
+// https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_fmadd&expand=2549
+FORCE_INLINE __m128 _mm_fmadd_ps(__m128 a, __m128 b, __m128 c)
```
Comment (danlark1) [inline sse2neon.h:2447]: Technically speaking, this is not an SSE extension, this was introduced with fma extension which took place even after AVX

File: tests/impl.cpp (hunk 6 above)
Comment (jserv) [reply tests/impl.cpp:989]: Continuous work on #81

**Pull request merged**